public/js/modules.js

```diff
@@ -576,6 +576,112 @@ SevenSegDisplay.prototype.customDraw = function() {
     ctx.stroke();
 }
 
+function SixteenSegDisplay(x, y, scope = globalScope) {
+    CircuitElement.call(this, x, y, scope, "RIGHT", 16);
+
+    this.directionFixed = true;
+    this.setDimensions(30, 50);
+
+      this.input1 = new Node( 0, -50, 0, this, bitwidth=16);
+      this.direction ="RIGHT";
+      this.dot = new Node( 0, 50, 0, this, bitwidth=1);
+}
+
+SixteenSegDisplay.prototype = Object.create(CircuitElement.prototype);
+SixteenSegDisplay.prototype.constructor = SixteenSegDisplay;
+SixteenSegDisplay.prototype.tooltipText = "Sixteen Display ToolTip: Consists of 16+1 bit inputs.";
+SixteenSegDisplay.prototype.customSave = function() {
+    var data = {
+
+        nodes: {
+            input1: findNode(this.input1),
+            dot: findNode(this.dot)    
+        }
+    }
+    return data;
+}
+
+SixteenSegDisplay.prototype.customDrawSegment = function(x1, y1, x2, y2, color) {
+    if (color == undefined) color = "lightgrey";
+    ctx = simulationArea.context;
+    ctx.beginPath();
+    ctx.strokeStyle = color;
+    ctx.lineWidth = correctWidth(4);
+    xx = this.x;
+    yy = this.y;
+    moveTo(ctx, x1, y1, xx, yy, this.direction);
+    lineTo(ctx, x2, y2, xx, yy, this.direction);
+    ctx.closePath();
+    ctx.stroke();
+}
+
+SixteenSegDisplay.prototype.customDrawSegmentSlant = function(x1, y1, x2, y2, color) {
+    if (color == undefined) color = "lightgrey";
+    ctx = simulationArea.context;
+    ctx.beginPath();
+    ctx.strokeStyle = color;
+    ctx.lineWidth = correctWidth(3);
+    xx = this.x;
+    yy = this.y;
+    moveTo(ctx, x1, y1, xx, yy, this.direction);
+    lineTo(ctx, x2, y2, xx, yy, this.direction);
+    ctx.closePath();
+    ctx.stroke();
+}
+
+SixteenSegDisplay.prototype.customDraw = function() {
```

Comment:
One optimization idea, just to keep the code more compact and slightly faster, would be to define an array with the coordinates and just extract the bits using `this.input.value & (1<<n)` (undefined coerces to zero automatically). This would avoid converting the number to string, then splitting it, then mapping to number, then assigning to multiple variables.

I might look like this:
```js
var value = this.input.value;
var color = ["lightgrey", "red"];
[
    [-20, -38, 0, -38],
    [20, -38, 0, -38],
    ... (and the others as you have in the calls to customDrawSegment & customDrawSegmentSlant)
].forEach(function (c,i){
if (i<12) {
     this.customDrawSegment(c[0],c[1],c[2],c[3], color[value & (1 << i)]);
} else {
    this.customDrawSegmentSlant(...same here...)
}
});
```

Reply:
If you do not want to use a loop it is fine, the code is still readable. I think it is a good idea to use a bitwise operation to perform slight optimization. However considering the amount of rendering we are doing, this hardly seems an overhead. Moreover, dec2bin internally uses `.toString()` which probably uses bitwise operations. However, this line seems unnecessary as you can directly index on the string and use that instead. 
`var arr = (binNum).toString(10).split("").map(Number);` as `binNum` is a string and can already be  accessed via index.

Reply:
@fpintos Actually the expression `[value & (1 << i)]` returns a value in decimal format and we need either a 0 or 1 to  get the correct results. Can you please guide me, how should I proceed.

Reply:
Sorry...make it `color[value & (1<<i) != 0]` and it should work.

Reply:
Or `color[ (value>>i) & 1 ]` could work as well ?